analytics/graph_record.go

```diff
@@ -0,0 +1,317 @@
+package analytics
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/TykTechnologies/graphql-go-tools/pkg/ast"
+	"github.com/TykTechnologies/graphql-go-tools/pkg/astnormalization"
+	"github.com/TykTechnologies/graphql-go-tools/pkg/astparser"
+	gql "github.com/TykTechnologies/graphql-go-tools/pkg/graphql"
+	"github.com/TykTechnologies/graphql-go-tools/pkg/operationreport"
+	"github.com/buger/jsonparser"
+)
+
+type GraphRecord struct {
+	Types map[string][]string
+
+	AnalyticsRecord `bson:",inline"`
+
+	OperationType string
+	Variables     string
+	Errors        []graphError
+	HasErrors     bool
+}
+
+func (a *AnalyticsRecord) ToGraphRecord() (GraphRecord, error) {
+	record := GraphRecord{
+		AnalyticsRecord: *a,
+	}
+	if a.ResponseCode >= 400 {
+		record.HasErrors = true
+	}
+	rawRequest, err := base64.StdEncoding.DecodeString(a.RawRequest)
+	if err != nil {
+		return record, fmt.Errorf("error decoding raw request: %w", err)
+	}
+
+	schemaBody, err := base64.StdEncoding.DecodeString(a.ApiSchema)
+	if err != nil {
+		return record, fmt.Errorf("error decoding schema: %w", err)
+	}
+
+	request, schema, operationName, err := generateNormalizedDocuments(rawRequest, schemaBody)
+	if err != nil {
+		return record, fmt.Errorf("error generating documents: %w", err)
+	}
+	if len(request.Input.Variables) != 0 && string(request.Input.Variables) != "null" {
+		record.Variables = base64.StdEncoding.EncodeToString(request.Input.Variables)
+	}
+
+	// get the operation ref
+	operationRef := 0
+	if operationName != "" {
+		for i := range request.OperationDefinitions {
+			if request.OperationDefinitionNameString(i) == operationName {
+				operationRef = i
+				break
+			}
+		}
+	} else if len(request.OperationDefinitions) > 1 {
+		return record, errors.New("no operation name specified")
+	}
+
+	// get operation type
+	switch request.OperationDefinitions[operationRef].OperationType {
+	case ast.OperationTypeMutation:
+		record.OperationType = string(ast.DefaultMutationTypeName)
+	case ast.OperationTypeSubscription:
+		record.OperationType = string(ast.DefaultSubscriptionTypeName)
+	case ast.OperationTypeQuery:
+		record.OperationType = string(ast.DefaultQueryTypeName)
+	}
+
+	// get the selection set types to start with
+	fieldTypeList, err := extractTypesOfSelectionSet(operationRef, request, schema)
+	if err != nil {
+		log.WithError(err).Error("error extracting selection set types")
+		return record, err
+	}
+	typesToFieldsMap := make(map[string][]string)
+	for fieldRef, typeDefRef := range fieldTypeList {
+		recursivelyExtractTypesAndFields(fieldRef, typeDefRef, typesToFieldsMap, request, schema)
+	}
+	record.Types = typesToFieldsMap
+
+	// get errors
+	responseDecoded, err := base64.StdEncoding.DecodeString(a.RawResponse)
+	if err != nil {
+		return record, nil
+	}
+	resp, err := http.ReadResponse(bufio.NewReader(bytes.NewReader(responseDecoded)), nil)
```

Comment:
@kofoworola why are we doing this? only to see if there's an error in the response body?

Reply:
Yes we need to read the response body properly to check for errors in the response so i used http's native read response to accommodate for different chunked response lengths etc